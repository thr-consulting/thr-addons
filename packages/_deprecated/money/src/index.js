import getCountryCode from './currencyCountry';
import GraphQLMoney from './graphql/GraphQLMoney';
import {makeMoney, formatMoney, roundTo, transformObjectsToMoney} from './util';

export {
	getCountryCode,
	GraphQLMoney,
	makeMoney,
	formatMoney,
	roundTo,
	transformObjectsToMoney,
};

/**
 * A JS-Money object.
 * @typedef Money
 */
