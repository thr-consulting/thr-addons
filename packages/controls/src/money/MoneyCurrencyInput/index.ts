export {MoneyCurrencyInput} from './MoneyCurrencyInput';
export type {MoneyCurrencyInputProps} from './MoneyCurrencyInput';
