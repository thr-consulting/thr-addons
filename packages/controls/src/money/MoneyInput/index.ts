export {MoneyInput} from './MoneyInput';
export type {MoneyInputProps} from './MoneyInput';
