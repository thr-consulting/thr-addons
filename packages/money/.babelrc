{
	"presets": [
		["@imperium/babel-preset-imperium", {"forceModules": true}]
	]
}
