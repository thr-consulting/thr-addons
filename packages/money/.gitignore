node_modules
lib
server.js
