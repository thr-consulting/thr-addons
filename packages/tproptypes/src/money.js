// @flow

import PropTypes from 'prop-types';
import Money from 'js-money';

/**
 * JS Money object
 */
export const money = PropTypes.instanceOf(Money);
