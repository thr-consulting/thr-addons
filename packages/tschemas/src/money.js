// @flow

import {mixed} from 'yup';
import Money from 'js-money';
import isNull from 'lodash/isNull';

/**
 * Validates a Money (from js-money).
 */
class MoneySchema extends mixed {
	constructor() {
		super({type: 'moneyschema'});

		this.transforms.push(value => {
			if (this.isType(value)) return value;

			if (isNull(value)) return null;
			if (value.amount && value.currency) return new Money(value.amount, value.currency);
			return null;
		});
	}

	_typeCheck(v: Object) {
		return v instanceof Money;
	}
}

export const money = () => new MoneySchema();
