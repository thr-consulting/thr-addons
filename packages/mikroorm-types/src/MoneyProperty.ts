import {Type, ValidationError} from '@mikro-orm/core';
import {isMoneyObject, toMoney} from '@thx/money';

export class MoneyProperty extends Type {
	convertToDatabaseValue(value: any): any {
		try {
			if (!value) return value;
			if (!isMoneyObject(value)) throw new Error();
			return JSON.stringify(value); // todo make sure this works properly
		} catch (err) {
			throw ValidationError.invalidType(MoneyProperty, value, 'JS');
		}
	}

	convertToJSValue(value: any): any {
		try {
			if (!value) return value;
			return toMoney(value);
		} catch (err) {
			throw ValidationError.invalidType(MoneyProperty, value, 'database');
		}
	}

	getColumnType() {
		return 'jsonb';
	}
}
