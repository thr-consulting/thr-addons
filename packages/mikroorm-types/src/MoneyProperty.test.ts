import {toMoney} from '@thx/money';
import {MoneyProperty} from './MoneyProperty';

describe('MoneyProperty', () => {
	it('should convert Money values', () => {
		const money = toMoney(100);

		const prop = new MoneyProperty();
		expect(prop.convertToDatabaseValue(money)).toMatchSnapshot();
		expect(prop.convertToJSValue(money)).toMatchSnapshot();
		expect(prop.getColumnType()).toMatchSnapshot();
	});

	it('should throw errors on invalid values', () => {
		const money = 'not money';
		const prop = new MoneyProperty();

		function convertDbValue() {
			prop.convertToDatabaseValue(money);
		}
		function convertJsValue() {
			prop.convertToJSValue(money);
		}

		expect(convertDbValue).toThrowErrorMatchingSnapshot();
		expect(convertJsValue).toThrowErrorMatchingSnapshot();
	});
});
