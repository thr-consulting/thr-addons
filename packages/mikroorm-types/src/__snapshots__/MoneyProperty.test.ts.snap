// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoneyProperty should convert Money values 1`] = `"{\\"amount\\":10000,\\"currency\\":\\"CAD\\"}"`;

exports[`MoneyProperty should convert Money values 2`] = `
Object {
  "amount": 10000,
  "currency": "CAD",
}
`;

exports[`MoneyProperty should convert Money values 3`] = `"jsonb"`;

exports[`MoneyProperty should throw errors on invalid values 1`] = `"Could not convert JS value 'not money' of type 'string' to type MoneyProperty"`;

exports[`MoneyProperty should throw errors on invalid values 2`] = `"Could not convert database value 'not money' of type 'string' to type MoneyProperty"`;
